cirq-core/cirq/protocols/apply_unitary_protocol.py

```diff
@@ -410,17 +422,18 @@ def _strat_apply_unitary_from_apply_unitary(
 
 
 def _strat_apply_unitary_from_unitary(
-    unitary_value: Any, args: ApplyUnitaryArgs
+    unitary_value: Any, args: ApplyUnitaryArgs, matrix: Optional[np.ndarray] = None
 ) -> Optional[np.ndarray]:
-    # Check for magic method.
-    method = getattr(unitary_value, '_unitary_', None)
-    if method is None:
-        return NotImplemented
+    if matrix is None:
+        # Check for magic method.
+        method = getattr(unitary_value, '_unitary_', None)
+        if method is None:
+            return NotImplemented
```

Comment:
Why is this needed? Looks like you are calling this method with the `matrix` from `_strat_apply_unitary_from_decompose` after you have already computed the narrowed matrix. Why the fallback on  `_strat_apply_unitary_from_decompose` ?